internal/transport/transport_test.go

```diff
@@ -2656,10 +2658,89 @@ func TestConnectionError_Unwrap(t *testing.T) {
 	}
 }
 
-// Test that in the event of a graceful client transport shutdown, i.e.,
+// TestClientSendsAGoAwayFrame verifies that in the event of a graceful client transport shutdown, i.e.,
 // clientTransport.Close(), client sends a goaway to the server with the correct
 // error code and debug data.
 func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	createClientServerConn(t, ConnectOptions{})
+}
+
+// writeHangSignal is used to hang the net.Conn Write for complete test duration.
+var writeHangSignal chan struct{}
+
+// hangingConn is a net.Conn wrapper for testing, simulating hanging connections
+// after a GOAWAY frame is sent, of which Write operations pause until explicitly signaled
+// or a timeout occurs.
+type hangingConn struct {
+	net.Conn
+}
+
+func (hc *hangingConn) Read(b []byte) (n int, err error) {
+	n, err = hc.Conn.Read(b)
+	return n, err
+}
+
+func (hc *hangingConn) Write(b []byte) (n int, err error) {
+	n, err = hc.Conn.Write(b)
+	if n == goAwayFrameSize { // GOAWAY frame
+		<-writeHangSignal
+	}
+	return n, err
+}
+
+func (hc *hangingConn) Close() error {
+	return hc.Conn.Close()
+}
+
+func (hc *hangingConn) LocalAddr() net.Addr {
+	return hc.Conn.LocalAddr()
+}
+
+func (hc *hangingConn) RemoteAddr() net.Addr {
+	return hc.Conn.RemoteAddr()
+}
+
+func (hc *hangingConn) SetDeadline(t time.Time) error {
+	return hc.Conn.SetDeadline(t)
+}
+
+func (hc *hangingConn) SetReadDeadline(t time.Time) error {
+	return hc.Conn.SetReadDeadline(t)
+}
+
+func (hc *hangingConn) SetWriteDeadline(t time.Time) error {
+	return hc.Conn.SetWriteDeadline(t)
+}
+
+func hangingDialer(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &hangingConn{Conn: conn}, nil
+}
+
+// TestClientCloseTimeoutOnHang verifies that in the event of a graceful
+// client transport shutdown, i.e., clientTransport.Close(), if the conn hung
+// for LoopyWriterTimeout, client should still be close itself and should
+// not wait for long.
+func (s) TestClientCloseTimeoutOnHang(t *testing.T) {
+	origGoAwayLoopyTimeout := GoAwayLoopyWriterTimeout
+	GoAwayLoopyWriterTimeout = 0
```

Comment:
If we are setting time out to 0, we probably don't need hanging conn as it will always time out? Can you try without hanging conn and see if time out still happens? If yes, test will become much more simpler